apps/server/src/modules/sync/events/events.gateway.ts

```diff
@@ -0,0 +1,83 @@
+import { Storage } from '@affine/storage';
+import { Inject } from '@nestjs/common';
+import {
+  ConnectedSocket,
+  MessageBody,
+  SubscribeMessage,
+  WebSocketGateway,
+  WebSocketServer,
+} from '@nestjs/websockets';
+import { Socket } from 'socket.io';
+
+import { StorageProvide } from '../../../storage';
+import { base64ToUint8Array, uint8ArrayToBase64 } from '../utils';
+import { WorkspaceService } from './workspace';
+
+const port = parseInt(process.env.PORT ?? '3010');
+
+@WebSocketGateway(port)
+export class EventsGateway {
+  constructor(
+    private readonly storageService: WorkspaceService,
+    @Inject(StorageProvide) private readonly storage: Storage
+  ) {}
+
+  @WebSocketServer()
+  server: any;
+
+  @SubscribeMessage('client-handshake')
+  async handleClientHandShake(
+    @MessageBody() workspace_id: string,
+    @ConnectedSocket() client: Socket
+  ) {
+    const docs = await this.storageService.getDocsFromWorkspaceId(workspace_id);
+    await client.join(workspace_id);
+
+    for (const { guid, update } of docs) {
+      this.server.to(workspace_id).emit('server-handshake', {
+        guid,
+        update: uint8ArrayToBase64(update),
+      });
+    }
+  }
+
+  @SubscribeMessage('client-update')
+  async handleClientUpdate(
+    @MessageBody()
+    message: {
+      workspaceId: string;
+      guid: string;
+      update: string;
+    }
+  ) {
+    const update = base64ToUint8Array(message.update);
```

Comment:
What's the purpose to convert `update` like this instead of just a simple `Buffer.from(message.update, 'base64')?

`base64 string -> Buffer -> Uint8Array -> Buffer`, the data is copied 3 times

Reply:
btw, `Buffer` is extended from `Uint8Array`, you can pass `Buffer` into any function that receives `Uint8Array`

Reply:
There's no particular purpose, I haven't noticed the cost, it's inlined now